compiler-plugin/src/main/kotlin/arrow/meta/quotes/ClassOrObject.kt

```diff
@@ -13,12 +13,112 @@ import org.jetbrains.kotlin.psi.psiUtil.getValueParameters
 import org.jetbrains.kotlin.psi.psiUtil.modalityModifierType
 import org.jetbrains.kotlin.psi.psiUtil.visibilityModifierType
 
+/**
+ * [classOrObject] is a function that intercepts all [KtClass] elements that [match]
+ * allowing a [Transform] to change the intercepted AST tree before compilation.
+ *
+ * An extension function of [Meta] and inheriting from [ExtensionPhase], [classOrObject] was designed to to feed in
+ * virtually any kind of [KtClass] predicate, followed by mapping a function that takes the desired [Scope] of our
+ * [KtClass] to change whatever PSI elements desired.
+ *
+ * For example, the [LensPlugin] and the [HigherKindPlugin] favor easy update of immutable data structures and ad-hoc polymorphism that does not require inheritance.
+ *
+ * In the [LensPlugin], we check to see if a type is a product type.  To add context, the cardinality of a product
+ * type is the product of its contained properties:
+ *
+ * ```
+ *      |(a, b)| = |a| x |b|
+ * ```
+ *
+ * In particular, the [LensPlugin] is passing a function to whatever class or object type is a product type:
+ *
+ * ```
+ *  val Meta.lenses: Plugin
+ *    get() =
+ *      "lenses" {
+ *        meta {
+ *          classOrObject(::isProductOrType) { c -> ... }
+ *        }
+ *      }
+ * ```
+ *
+ * Where the function type for [::isProductType] is written as:
+ *
+ * ```
+ * fun isProductType(ktClass: KtClass): Boolean =
+ *      ktClass.isData() &&
+ *        ktClass.primaryConstructorParameters.isNotEmpty() &&
+ *        ktClass.primaryConstructorParameters.all { !it.isMutable } &&
+ *        ktClass.typeParameters.isEmpty()
+ * ```
+ *
+ * For a [KtClass] to be a product type, we check to see if the [KtClass] is an ADT, and that the cardinality of
```

Comment:
nit: "is an ADT" -> is that accurate? as in: does `isData()` check that or just if it's a `data class` ? i.e: sealed classes and enums are considered sum types hence ADTs if I'm not mistaken. In case `isData()` does not check for those I think I'd update the wording here. Otherwise please ignore me.

Reply:
Let's double check with @raulraja but I believe that the proof behind a product type lays in algebraic data types - the proof doesn't line up exactly with the code, but I think it's close with this idea:

The cardinality of a product type is the product of their cardinalities. 

    |(a, b)| = |a| x |b|

Reply:
Maybe I'm being too didactic, but I'm very new about thinking with types! So this could easily be wrong

Reply:
Yep that's right but I think the wording is wrong, and maybe the implementation of the plugin too. Given its the implementation of `isProductType()`, we need to check for data classes and also any other aggregation types like all the types of tuples. All those are product types. I think the implementation is just checking for data classes here, so it'd be incomplete. Also, the wording in the description shouldn't refer to ADTs overall since that also implies checking for sum types which is not correct for this one.

Reply:
Thank you so much for the feedback, that's incredibly helpful! but also lol... that's a good point. This is @nomisRev's plugin that we're looking at I think with LensPlugin - perhaps the productType check is modified for a specific purpose for LensPlugin?